sstable_set: maintain bytes on disk

bhalevy opened the pull request:
and use that in compaction_group, rather than
respective accumulators of its own.

This is part of of larger series to make cache updates exception safe.

Refs #14043

avikivity commented on sstables/sstable_set.cc:
```diff
@@ -172,14 +172,14 @@ stop_iteration sstable_set::for_each_sstable_until(std::function<stop_iteration(
+bool
 sstable_set::erase(shared_sstable sst) {
-    _impl->erase(sst);
+    return _impl->erase(sst);
```
I understand how this came about, but can't we prevent it? This will infect the API.

I can't think of a way to prevent it. If we had on_internal_error() since the beginning it would be possible, but adding them now is difficult.

bhalevy replied:
Right.  I don't think it's a common case, but I can't say for sure erase would only be called for sstables that exist in the set.

raphaelsc commented:
bytes_on_disk would be incorrect for compound sets, where each individual erase and insertion will go through each individual set. I am planning to decouple compound sets (another user of compound set is tablet aware one) and individual sstable set, as they don't even converge in what's needed from the interface.

bhalevy commented:
> bytes_on_disk would be incorrect for compound sets, where each individual erase and insertion will go through each individual set. I am planning to decouple compound sets (another user of compound set is tablet aware one) and individual sstable set, as they don't even converge in what's needed from the interface.

I don't think it's a problem because `compound_sstable_set::insert` and `erase` are defunct, and throw `std::bad_function_call`.  We do the sstable size bookkeeping in the underlying sstable sets.

avikivity commented:
> > bytes_on_disk would be incorrect for compound sets, where each individual erase and insertion will go through each individual set. I am planning to decouple compound sets (another user of compound set is tablet aware one) and individual sstable set, as they don't even converge in what's needed from the interface.
> 
> I don't think it's a problem because `compound_sstable_set::insert` and `erase` are defunct, and throw `std::bad_function_call`. We do the sstable size bookkeeping in the underlying sstable sets.

But then the compound set's bytes_on_disk would be zero, no?

bhalevy commented:
> > > bytes_on_disk would be incorrect for compound sets, where each individual erase and insertion will go through each individual set. I am planning to decouple compound sets (another user of compound set is tablet aware one) and individual sstable set, as they don't even converge in what's needed from the interface.
> > 
> > 
> > I don't think it's a problem because `compound_sstable_set::insert` and `erase` are defunct, and throw `std::bad_function_call`. We do the sstable size bookkeeping in the underlying sstable sets.
> 
> But then the compound set's bytes_on_disk would be zero, no?

We never check that on the compound set.
If we care about that enough, we can define the bytes_on_disk() method as virtual and override it in the compound sstable set, where we can return the sum of all underlying sstable sets.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3061/

bhalevy commented:
In v2 ([f54ab48](https://github.com/scylladb/scylladb/commit/f54ab4827334373dfa203a0285fb090d212654b5)):
- moved implementation of bookkeeping to the sstable_set_impl
- override `bytes_on_disk` by `compound_sstable_set`, return sum of all underlying sets.
- added a couple unit tests for `partitioned_sstable_set` and `time_series_sstable_set`
  - both test also compound_sstable_set of the former set impl.

raphaelsc commented:
now that we moved the tracking of bytes on disk to sstable_set_impl, do we still need the first patch?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3065/

avikivity commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3065/

https://github.com/scylladb/scylla-dtest/pull/3502 ?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3070/
